src/cpp/parse.cpp

```diff
@@ -473,12 +473,19 @@ void Parser::merge_types(const typet &src, typet &dest)
   {
     if(dest.id()!=ID_merged_type)
     {
+      source_locationt location=dest.source_location();
       typet tmp(ID_merged_type);
       tmp.move_to_subtypes(dest);
+      tmp.add_source_location()=location;
       dest=tmp;
     }
 
-    dest.copy_to_subtypes(src);
+    // the end of the subtypes container needs to stay the same,
+    // since several analysis functions traverse via the end for
+    // merged_types
+    typet::subtypest &sub=dest.subtypes();
+    sub.emplace(sub.begin(), src);
```

Comment:
Might as well use insert, since emplace is just calling a copy-constructor itself here plus doing some mildly-costly (in my experience) gymnastics to avoid copying the argument in the meantime